### PR TITLE
Added resumable upload

### PR DESCRIPTION
**Pull request by mohami2000**
https://b.corp.google.com/issues/312511716

### CODE REVIEW

File: src/kagglehub/gcs_upload.py [hunk 1]
```diff
@@ -72,18 +101,33 @@ def _upload_blob(file_path: str, model_type: str):
...
+                if upload_response.status_code in [200, 201]:
+                    return response["token"]
+                elif upload_response.status_code == 308:  # Resume Incomplete # noqa: PLR2004
+                    uploaded_bytes = _check_uploaded_size(session_uri, file_size)
+                else:
+                    upload_failed_exception = f"Upload failed: {upload_response.text}"
+                    raise Exception(upload_failed_exception)
+            except (requests.ConnectionError, requests.Timeout) as e:
+                logger.info(f"Network issue: {e}, retrying...")
+                retry_count += 1
```
Comment (rosbo) [inline src/kagglehub/gcs_upload.py:129]: Should we add a sleep here?

File: src/kagglehub/gcs_upload.py [hunk 2]
```diff
@@ -72,18 +101,33 @@ def _upload_blob(file_path: str, model_type: str):
...
+                headers["Content-Range"] = f"bytes {uploaded_bytes}-{file_size - 1}/{file_size}"
+                upload_response = requests.put(
+                    session_uri, headers=headers, data=reader_wrapper, timeout=REQUEST_TIMEOUT
                 )
-                raise BackendError(upload_fail_message)
+
+                if upload_response.status_code in [200, 201]:
+                    return response["token"]
+                elif upload_response.status_code == 308:  # Resume Incomplete # noqa: PLR2004
+                    uploaded_bytes = _check_uploaded_size(session_uri, file_size)
```
Comment (rosbo) [inline src/kagglehub/gcs_upload.py:123]: Can you add a test for this case?

File: src/kagglehub/gcs_upload.py [hunk 3]
```diff
@@ -72,18 +101,33 @@ def _upload_blob(file_path: str, model_type: str):
...
+                else:
+                    upload_failed_exception = f"Upload failed: {upload_response.text}"
+                    raise Exception(upload_failed_exception)
+            except (requests.ConnectionError, requests.Timeout) as e:
```
Comment (rosbo) [inline src/kagglehub/gcs_upload.py:127]: Should we retry on Timeout? Or only ConnectionError?

File: src/kagglehub/gcs_upload.py [hunk 4]
```diff
@@ -72,18 +101,33 @@ def _upload_blob(file_path: str, model_type: str):
...
                 )
-                raise BackendError(upload_fail_message)
+
+                if upload_response.status_code in [200, 201]:
+                    return response["token"]
+                elif upload_response.status_code == 308:  # Resume Incomplete # noqa: PLR2004
+                    uploaded_bytes = _check_uploaded_size(session_uri, file_size)
+                else:
+                    upload_failed_exception = f"Upload failed: {upload_response.text}"
+                    raise Exception(upload_failed_exception)
```
Comment (rosbo) [inline src/kagglehub/gcs_upload.py:126]: Let's raise a BackendError instead and include the status code like before: https://github.com/Kaggle/kagglehub/blob/5df7847bb21116b7cc29a04af4d556ac8308716a/src/kagglehub/gcs_upload.py#L70

**Review by rosbo — CHANGES_REQUESTED** [review]

File: src/kagglehub/gcs_upload.py (hunk 3 above)
Comment (mohami2000) [reply src/kagglehub/gcs_upload.py:127]: I think both, right? they occur for diff reason so makes sense to cover them both. connection error is for connectivity issues and timeout happens if the the data is taking too long to transfer over a network or things like that

File: tests/test_model_upload.py [hunk 5]
```diff
@@ -84,14 +85,24 @@ def do_POST(self):  # noqa: N802
 
 
 class GcsAPIHandler(BaseHTTPRequestHandler):
-    def do_HEAD(self):  # noqa: N802
-        self.send_response(200)
+    simulate_308 = False
```
Comment (rosbo) [inline tests/test_model_upload.py:88]: Make sure to reset these two values in the `setUp` method.

**Review by rosbo — APPROVED** [review]
Great work!

**Pull request merged**